src/node/blockstorage.cpp

```diff
@@ -1072,23 +1072,31 @@ bool BlockManager::ReadRawBlock(std::vector<std::byte>& block, const FlatFilePos
         if (blk_start != GetParams().MessageStart()) {
             LogError("Block magic mismatch for %s: %s versus expected %s while reading raw block",
                 pos.ToString(), HexStr(blk_start), HexStr(GetParams().MessageStart()));
-            return false;
+            return util::Unexpected{ReadRawError::IO};
         }
 
         if (blk_size > MAX_SIZE) {
             LogError("Block data is larger than maximum deserialization size for %s: %s versus %s while reading raw block",
                 pos.ToString(), blk_size, MAX_SIZE);
-            return false;
+            return util::Unexpected{ReadRawError::IO};
+        }
+
+        if (block_part) {
+            const auto [offset, size]{*block_part};
+            if (size == 0 || offset >= blk_size || size > blk_size - offset) {
```

Comment:
```suggestion
            if (size == 0 || SaturatingAdd(offset, size) > blk_size) {
```

nit in the second commit: Could use a single saturating add and a single `>` compare?

Reply:
nit: Also seems fine to allow zero-size?

Reply:
what would be the advantage of zero size?

Reply:
> what would be the advantage of zero size?

idk. It is generally valid to read zero bytes:

```
with open("example.txt", "rb") as f:
    data = f.read(0)
    print(data)   # prints: b''
```

i don't mind either way. I just wanted to mention it, because some non-ideal hacky script may read blocks in ranges and read the last range, even if it is of zero size. yes, the script should be adjusted, but I don't see the harm in allowing it. In any case, just a nit and not important at all.

Reply:
nit: Agree on reducing the number of conditions.
```suggestion
            if (size == 0 || SaturatingAdd(offset, size) >= blk_size) {
```
No strong opinion for/against disallowing 0-size ranges.